File: src/bumperbot_cpp_examples/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include <chrono>

using namespace std::chrono_literals; // to be able to use 1s literal

// Define the SimplePublisher class as a subclass of rclcpp::Node
class SimplePublisher : public rclcpp::Node
{
public:
    // constructor, initialize Node with node name "simple_publisher" and counter to 0
    SimplePublisher() : Node("simple_publisher"), counter_(0)
    {
        // Create a publisher on the "chatter" topic
        pub_ = create_publisher<std_msgs::msg::String>(
            "chatter", // topic name
            10         // size of queue
        );

        // create a timer to launch callback at 1s intervals
        timer_ = create_wall_timer(
            1s,                                               // interval to call the callback function (literal: 1 second)
            std::bind(&SimplePublisher::timerCallback, this) // callback function
        );

        RCLCPP_INFO(get_logger(), "Publishing at 1 Hz");
    }

private:
    unsigned int counter_;                                    // counter to keep track of the number of messages published
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_; // shared pointer to the publisher
    rclcpp::TimerBase::SharedPtr timer_;                      // shared pointer to the timer

    void timerCallback() // callback function to be called at 1s intervals
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello ROS2! [" + std::to_string(counter_++) + "]";

        pub_->publish(message);
    }
};


int main (int argc, char* argv[])
{
    rclcpp::init(argc, argv); // initialize ROS2

    auto node = std::make_shared<SimplePublisher>(); // create a shared pointer to the SimplePublisher class
    rclcpp::spin(node); // run the node

    rclcpp::shutdown(); // shutdown ROS2

    return 0;
}

File: src/bumperbot_py_examples/bumperbot_py_examples/simple_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String

# create a publisher class
class SimplePublisher(Node): # inherit from Node
    
    def __init__(self): # constructor
        
        # call constructor of the base class
        super().__init__('simple_publisher') 
        
        # create a publisher for String messages in channel chatter, 
        # max size 10
        self.pub_ = self.create_publisher(String, 'chatter', 10) 
        
        self.counter_ = 0
        self.frequency_ = 1.0
        
        # console output
        self.get_logger().info('Publishing at %d Hz' % self.frequency_)

        # create timer object: executes a callback at regular times interval.
        self.timer = self.create_timer(self.frequency_, self.timerCallback)

    def timerCallback(self): #define callback as a member of the class
        msg = String() # create message
        msg.data = 'Hello, ROS2 - counter: %d ' % self.counter_

        self.pub_.publish(msg) # start publishing
        self.counter_ += 1 # increment counter

def main():

    # 1) Initialization: calling `init` for a particular Context. 
    # Must be done before any ROS nodes can be created.
    rclpy.init()
    

    # 2) Create one or more ROS nodes: by calling `create_node` or instantiating
    # a Node. Nodes may implement common ROS entities such as publishers, 
    # subscriptions, services and actions.
    simple_publisher = SimplePublisher() # create a SimplePublisher object

    # 3) Process node callbacks: by calling `spin`, `spin_once`, or 
    # `spin_until_future_complete`.
    # After a node is created, one can process work items (e.g. subscription 
    # callbacks) that are waiting to be executed
    rclpy.spin(simple_publisher) # process callbacks

    # 4) Shutdown
    # When finished with a previously initialized .Context (i.e. done using
    # all ROS nodes associated with the context), the shutdown function should 
    # be called. This will invalidate all entities derived from the context.
    simple_publisher.destroy_node() # destroy node
    rclpy.shutdown() # shutdown ROS2

if __name__ == '__main__':
    main()